handout/routing-advanced/route_guards.md

```diff
@@ -0,0 +1,84 @@
+# Route Guards #
+
+Route guards provide a way to control whether the user can navigate to or navigate away from a given route.
+
+## CanActivate ##
+
+Some parts of the application may only be accessible under certain conditions. For example, routes may only be accessible once the user has logged in or accepted Terms & Conditions. We can use route guards to check these conditions and control access to routes.
+
+Let's look at an example activate guard that checks whether the user is logged in:
+
+```javscript
+import { CanActivate } from '@angular/router';
+import { Injectable } from '@angular/core';
+import { LoginService } from './login-service';
+
+@Injectable()
+export class LoginRouteGuard implements CanActivate {
+
+  constructor(private loginService: LoginService) {}
+
+  canActivate() {
+    return this.loginService.isLoggedIn();
+  }
+}
+```
+
+This class implements the `CanActivate` interface by implementing the `canActivate` function.
+
+When `canActivate` returns true, the user can activate the route. When `canActivate` returns false, the user cannot access the route. In the above example, we allow access when the user is logged in.
+
+We can do other things in the `canActivate` function as well such as notify the user that they can't access that part of the application, or redirect them to the login page.
+
+[See Official Definition for CanActivate](https://angular.io/docs/ts/latest/api/router/index/CanActivate-interface.html)
+
+## CanDeactivate ##
+
+Route guards can also control whether a user can leave a certain route.
+
+For example, say the user has typed information into a form on the page, but has not submitted the form. If they were to leave the page, they would lose the information. We may want to prompt the user if the user attempts to leave the route without submitting or saving the information.
+
+This works the exact same way as `CanActivate` with a `canDeactivate` on the `CanDeactivate` interface instead.
+
+[See Official Definition for CanDeactivate](https://angular.io/docs/ts/latest/api/router/index/CanDeactivate-interface.html)
+
+## Async route guards ##
+
+The `canActivate` and `canDeactivate` functions can either return values of type `boolean`, or `Observable<boolean>` (an Observable that resolve to `boolean`). If you need to do an asynchronous request (like a server request) to determine whether the user can navigate to or away from the route, you can simply return an `Observable<boolean>`. The router will wait until it is resolved and use that value to determine access.
+
+## Registering the route guards with routes ##
```

Comment:
I think the flow of this could be improved a little.  I think putting this section up top makes more sense, so the flow of reading is

* What is a guard, and their use cases
* Registering a guard with a route
* Implementing a canActivate guard
* Implementing a canDeactivate guard - could simply do a 'return confirm('Are you sure?')